### ISSUE TITLE
Question:  how does the playerId localStorage work?

### ISSUE THREAD

**Issue by joshua-a-becker** [empiricaly/meteor-empirica-core#29]
I'm trying to understand something and getting stuck just by reading the code, and not quite sure how to test it thoroughly enough to understand the possible edge cases.

Suppose I run an empirica app on a domain my-lab.meteorapp.com and then a few weeks later, run a completely different app (with a completely different database) on the exact same domain (my-lab.meteorapp.com)

What is the expected behavior if the first and second app are both accessed on the same computer with the same web brower, but by different people?  (I am assuming that the localStorage item in the web browser has not been cleared for any reason)

In short:  is this totally safe, or will the 2nd user will be assigned the playerId of the first user?

**Comment by npaton** [comment]
> I'm trying to understand something and getting stuck just by reading the code, and not quite sure how to test it thoroughly enough to understand the possible edge cases.
> 
> Suppose I run an empirica app on a domain my-lab.meteorapp.com and then a few weeks later, run a completely different app (with a completely different database) on the exact same domain (my-lab.meteorapp.com)

If the database is different, the player with payerId stored in localStorage will not be found, and the user will be treated as a new player (presented with the NewPlayer form or auto register if playerParamId is set).

> What is the expected behavior if the first and second app are both accessed on the same computer with the same web browser, but by different people? (I am assuming that the localStorage item in the web browser has not been cleared for any reason)

This is a similar case, but sounds like the people playing are different. Same behavior, the DB does not contain the old playerId, so it should work fine.

> In short: is this totally safe, or will the 2nd user will be assigned the playerId of the first user?

It should be. It is not something that was extensively tested, so I recommended trying it out, just to be sure.

For a very quick explanation of what's happening behind the curtains, we first try to load the player using the playerId we store in localStorage. This allows for people to come back to their game (for whatever reasons: browser crashed, clicked reload, finish game tomorrow, etc.)

If the DB is explicitly cleared (with the admin Reset buttons), the localStorage will be cleared (this is particularly useful during local development)

If, on the other hand, the player is not found with the playerId in the DB, we do not automatically clear the localStorage, but we do move onto the new player registration process, as we've confirmed not seeing the player. If/when the player does register again, we replace the playerId stored in localStorage.

In this last scenario, we could be clear the playerId stored in localStorage if we seem to not have found it in the DB. The only reason not to, is because we aren't 100% certain we haven't found it. We are just sure, we haven't received the player from the server. And maybe we couldn't receive the player from the server for other reasons, the simplest one being, we're not connected to the server (no internet or server down, for example). So our current behavior is conservative, but should still work fine with scenarios where the DB has voluntarily been cleared (such as replacing the experiment entirely).

Hope that all makes sense. If it does, and you are not having any issues in your tests, feel free to close the issue. ⛵️

**Comment by joshua-a-becker** [comment]
Thanks, this is the key piece I needed to know, highlighting for my own future reference:

> If, on the other hand, the player is not found with the playerId in the DB, we do not automatically clear the localStorage, but we do move onto the new player registration process, as we've confirmed not seeing the player. If/when the player does register again, we replace the playerId stored in localStorage.

i was also concerned about the case where the *same* user accesses again.  sounds like as long as it's a new/empty database (e.g., because i'm running a new experiment, in this hypothetical) then it will function as expected no matter who it is.

**Comment by joshua-a-becker** [comment]
**ah new question:**    suppose that the playerIdParam is supplied by the URL, and this conflicts with the playerIdParam in the LocalStorage:   which wins?

**Comment by npaton** [comment]
`playerParamId` and playerId storage in localStorage are cardinal features, they should not affect each other, in most cases (see exception below). In other words, nothing in the localStorage lookup is referencing playerParamId.

If you take a look at the source file for the playerId saving in [IdentifiedController](https://github.com/empiricaly/meteor-empirica-core/blob/master/ui/containers/IdentifiedContainer.jsx), you will not find any reference to `playerIdParam`. In [PublicContainer](https://github.com/empiricaly/meteor-empirica-core/blob/master/ui/containers/PublicContainer.jsx), where we handle the player auto-registration, nothing makes reference to `playerIdKey`.

You do bring up a potential problem (though somewhat unlikely at our scale), we do use a URL parameter for this feature (`playerIdKey`), and if you use `playerIdKey` as your `playerIdParam`, it could break thing. Maybe we should namespace that parameter better, to lower further the possibility of a conflict. I do not imagine at the moment `playerIdKey` is something people would generally use as their `playerIdParam` configuration value. We could maybe call this param something like `empirica_playerIdKey`.

Note that (and I haven't checked this) even in the case where `playerParamId` is set to `playerKeyId`, the features will overlap oddly, but they should actually still work. `playerKeyId` is just telling the app under what key to store the saved `playerId`, and while the `playerId` and the key with which it is stored in the localStorage will be the same, it should still work.

So all in all, it should just work. If this problem appears recurrent and problematic, we do have the option of moving to something like `empirica_playerIdKey` or `empirica_playerIdKey_please_dont_use_this_param` 😄. Feel free to create a ticket about that, let me know in your tests if my assumptions are correct, and close this issue if it's all working as expected. ⛵️

**Comment by joshua-a-becker** [comment]
All working as expected---I have not seen any problems, I was just trying to anticipate possible edge cases if, e.g., someone runs different experiments with the same URL.

I am thinking of the case where user A accesses the site, their playerId param is stored in localstorage.  THEN user B enters the site with the same browser and their playerId is given via URL, but they are still recognized as player A by the localStorage.

It looks like the localStorage will win, as per line 58 of [Public.jsx](https://github.com/empiricaly/meteor-empirica-core/blob/master/ui/components/Public.jsx) --- the player object is already filled due to finding the localStorage item and so NewPlayer.jsx is never called?

FWIW, I think this is *exceedingly unlikely* to happen---it would require two different humans to be using the same web browser to access the same experiment sequentially.
